### PR TITLE
Switch to use rdkafka client

### PR DESCRIPTION
**Pull request by bzwei**
Completely replace ruby-kafka with rdkafka client.
Bump version to 0.2.0

Replace #53

### CODE REVIEW

**Comment by bzwei** [comment]
@Fryguy There is no optional dependencies for the rdkafka gem, but it requires ruby 2.4. Therefore the build on 2.3.1 failed. Should we remove this version from Travis CI?

**Comment by Fryguy** [comment]
Yeah I'm ok with dropping Ruby 2.3 support.  We can cut a major version if we think it's a big deal.

**Comment by Fryguy** [comment]
cc @simaishi rdkafka would be a new C C extension gem.

File: .travis.yml
Comment (Fryguy) [inline .travis.yml:4]: Can you also add the latest 2.5 and 2.6 here?  Then we can be sure it works with newer rubies.

File: lib/manageiq/messaging/version.rb
Comment (Fryguy) [inline lib/manageiq/messaging/version.rb:3]: Let's not do thsi change in this PR...we should decide separately the version (more likely it's going to be major as I see something options being dropped, like the ssl/sasl stuff

File: lib/manageiq/messaging/kafka/queue.rb
Comment (Fryguy) [inline lib/manageiq/messaging/kafka/queue.rb:5]: Not sure I like this being manageiq specific, but it's probably ok for now (maybe we should make this a parameter option somewhere?)

File: lib/manageiq/messaging/kafka/queue.rb
Comment (Fryguy) [inline lib/manageiq/messaging/kafka/queue.rb:26]: minor, but you have a 4 space indent here.

File: lib/manageiq/messaging/kafka/topic.rb
Comment (Fryguy) [inline lib/manageiq/messaging/kafka/topic.rb:20]: 4 space indent

**Review by Fryguy — APPROVED** [review]
LGTM...would like @agrare to review.

File: lib/manageiq/messaging/client.rb
Comment (agrare) [inline lib/manageiq/messaging/client.rb:52]: @bzwei is this needed because of the switch to rdkafka or unrelated?  Seems like overkill when the caller could just put `client.close` in an `ensure` block and I don't know that I want a gem messing with my signal handlers.  What happens if the caller called `.close` already?

File: lib/manageiq/messaging/kafka/client.rb
Comment (agrare) [inline lib/manageiq/messaging/kafka/client.rb:66]: If the upstream is responsive maybe you could open a PR to add a `Rdkafka::Producer#list_topics` method so we don't have to `instance_variable_get` ?

File: lib/manageiq/messaging/client.rb
Comment (bzwei) [reply lib/manageiq/messaging/client.rb:52]: Not specifically for rdkafka. When we have a thread subscribe to a topic, the thread normally runs until the app shuts down. This code ensures the connection closes properly if the caller does not have a chance to do so. It is safe to call `.close` multiple times.

Because we do `old.call`,  existing signal handler is not overridden.

File: lib/manageiq/messaging/kafka/client.rb
Comment (bzwei) [reply lib/manageiq/messaging/kafka/client.rb:66]: Yes we could. I add this only because sources-api uses this feature now.

Comment (agrare) [reply lib/manageiq/messaging/kafka/client.rb:66]: Yeah it seems like a great feature to have

File: lib/manageiq/messaging/client.rb
Comment (agrare) [reply lib/manageiq/messaging/client.rb:52]: Yeah I realize it isn't overridden but seems like extra complexity when a simple `ensure` would do the same thing.  Anyway I won't hold this up for this just seems odd

Comment (agrare) [reply lib/manageiq/messaging/client.rb:52]: Actually an `ensure` would be better because this only cleans up the connection when the process exist where an ensure at the end of the method would close it immediately, e.g.

```
def worker_thread
  loop do
    client = ManageIQ::Messaging::Client.open
    client.subscribe_topic do
    end
  rescue => err
    # log an error and retry
  end
end
```

Would open _lots_ of connections without closing them

File: lib/manageiq/messaging/kafka/client.rb
Comment (agrare) [inline lib/manageiq/messaging/kafka/client.rb:52]: Is a `no_offset` error something that no client would ever want an exception for from `.commit`?

Comment (agrare) [reply lib/manageiq/messaging/kafka/client.rb:52]: > You can generally handle the NO_OFFSETS error as a success-indicator, the offsets are already committed.

From https://github.com/confluentinc/confluent-kafka-python/issues/295 so this looks fine then

**Review by agrare — APPROVED** [review]

File: lib/manageiq/messaging/kafka/client.rb
Comment (bzwei) [reply lib/manageiq/messaging/kafka/client.rb:52]: I found this error when we tried to ack a message multiple times. I think it should ignore this error.

File: lib/manageiq/messaging/client.rb
Comment (bzwei) [reply lib/manageiq/messaging/client.rb:52]: It is the responsibility of the caller who uses messaging gem to do `ensure`. Capturing `TERM` is an additional helper offered by this gem. But you're right. It should not be offered by a gem. I'll remove it.

**Review by agrare — APPROVED** [review]

**Pull request merged**
